src/security/context.cpp

```diff
@@ -45,6 +50,35 @@ namespace {
 
 namespace dnsec = datadog::nginx::security;
 
+std::size_t chain_length(ngx_chain_t const *ch) {
+  std::size_t len = 0;
+  for (ngx_chain_t const *cl = ch; cl; cl = cl->next) {
+    len++;
+  }
+  return len;
+}
+std::size_t chain_size(ngx_chain_t const *ch) {
+  std::size_t size = 0;
+  for (ngx_chain_t const *cl = ch; cl; cl = cl->next) {
+    size += ngx_buf_size(cl->buf);
+  }
+  return size;
+}
+std::size_t has_special(ngx_chain_t const *ch) {
+  for (ngx_chain_t const *cl = ch; cl; cl = cl->next) {
+    return ngx_buf_special(ch->buf);
```

Comment:
This doesn't look correct as you're returning on the first iteration and you're not even looking at `cl`.